### PR TITLE
Fix some resource leaks when running rpmbuild

### PR DESCRIPTION
**Pull request by JetXujing**
Fix some resource leaks in some cases when running rpmbuild.

### CODE REVIEW

**Comment by pmatilai** [comment]
> "emsg" is leaked in some cases when running rpmbuild, because "emsg"
may be alloced and be set when checkDep returns RPMRC_OK.

If emsg is allocated when checkDep() returns without an error, then *that* should be fixed instead I think. What's the case where that happens, I didn't see anything immediately obvious there?

**Comment by JetXujing** [comment]
> > "emsg" is leaked in some cases when running rpmbuild, because "emsg"
> > may be alloced and be set when checkDep returns RPMRC_OK.
> 
> If emsg is allocated when checkDep() returns without an error, then perhaps _that_ should be fixed instead I think. What's the case where that happens, I didn't see anything immediately obvious there?

When `checkSep(EVR, '-', emsg)` or `checkSep(EVR, ':', emsg)` or `checkEpoch(EVR, emsg)` returns RPMRC_FAIL, and `_wrong_version_format_terminate_build` is defined as 0, the checkDep() will returns RPMRC_OK and emsg is set.
```
if (checkSep(EVR, '-', emsg) != RPMRC_OK ||
	    checkSep(EVR, ':', emsg) != RPMRC_OK ||
	    checkEpoch(EVR, emsg) != RPMRC_OK) {

	    if (rpmExpandNumeric("%{?_wrong_version_format_terminate_build}"))
		return RPMRC_FAIL;
	}
```

**Comment by pmatilai** [comment]
Didn't look at the patches too closely yet, but:
- Please capitalize commit messages properly, including the summary line. As in, start sentences with a capital letter.
- Avoid piling on large number of commits to a single PR, it makes reviewing harder even if they're just tiny one-liners. 
- If something needs more than one or two frees, it needs a different approach entirely (eg the sbp leak here)

**Comment by JetXujing** [comment]
> Didn't look at the patches too closely yet, but:
> 
> * Please capitalize commit messages properly, including the summary line. As in, start sentences with a capital letter.
> * Avoid piling on large number of commits to a single PR, it makes reviewing harder even if they're just tiny one-liners. Not every one-liner needs a separate PR and there's no hard line, but this is certainly at the upper limit already.
> * If something needs more than one or two frees, it needs a different approach entirely (eg the sbp leak here)

The leaked scenarios are different and independent. Is it clearer to submit and fix a scenario separately?

File: build/build.c [hunk 1]
```diff
@@ -276,7 +279,11 @@ static int doBuildRequires(rpmSpec spec, int test)
 
  exit:
     freeStringBuf(sb_stdout);
-    free(output);
+    if (output) {
```
Comment (pmatilai) [inline build/build.c:282]: Good catch, but this should simply be `argvFree(output)` instead.

Comment (JetXujing) [reply build/build.c]: Thank you for the tip, I've made the relevant changes.

**Comment by JetXujing** [comment]
> > When `checkSep(EVR, '-', emsg)` or `checkSep(EVR, ':', emsg)` or `checkEpoch(EVR, emsg)` returns RPMRC_FAIL, and `_wrong_version_format_terminate_build` is defined as 0, the checkDep() will returns RPMRC_OK and emsg is set.
> 
> Oh, indeed. We could just remove _wrong_version_format_terminate_build, that has long since served its purpose, but then I spotted another case where this happens: the "it's not recommended to have..." warning messages.
> 
> In both these cases a warning will get swallowed in case of an error, that's not a sane behavior even if the leak was fixed. This will need something rather different, please drop that particular commit from this series to let us move forward with the simpler cases. (that's always the gotcha with multiple unrelated commits in a PR: any one of them can stop the whole thing)

I agree with you that even if the leak is fixed, there are still other problems, and I've deleted the corresponding submission,

**Comment by JetXujing** [comment]
Yes, I agree with you. When there are too many submissions in a PR, one problematic submission will block other submissions. Considering that the repair code is simple and is a leak problem of rpmbuild, I put them in a PR to prevent too many PRs.

**Comment by pmatilai** [comment]
The rest seems straightforward enough. Thanks for the patches!

**Pull request merged**